streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java

```diff
@@ -77,10 +77,10 @@ Segment getOrCreateSegment(final long segmentId, final ProcessorContext context)
             if (!isSegment(segment, segmentId)) {
                 cleanup(segmentId);
             }
-            if (!segments.containsKey(key)) {
-                Segment newSegment = new Segment(segmentName(segmentId), name, segmentId);
+            Segment newSegment = new Segment(segmentName(segmentId), name, segmentId);
+            Segment previousSegment = segments.putIfAbsent(key, newSegment);
+            if (previousSegment == null) {
```

Comment:
@ijuma Thanks for reviewing. Actually I thought so but since get operation has O(1) time complexity, so I did not make the change earlier.
I have made this change now. Kindly check once.